lib/elixir/lib/calendar/iso.ex

```diff
@@ -1455,6 +1518,231 @@ defmodule Calendar.ISO do
     {days, {@parts_per_day - 1, @parts_per_day}}
   end
 
+  @doc """
+  Shifts date by Duration units according to its calendar.
+
+  Available units are: `:year, :month, :week, :day, :hour, :minute, :second, :microsecond`.
+
+  When shifting by month:
+  - it will shift to the current day of a month
+  - when the current day does not exist in a month, it will shift to the last day of a month
+
+  ## Examples
+
+      iex> Calendar.ISO.shift_date(2016, 1, 3, Duration.new!(month: 2))
+      {2016, 3, 3}
+      iex> Calendar.ISO.shift_date(2016, 2, 29, Duration.new!(month: 1))
+      {2016, 3, 29}
+      iex> Calendar.ISO.shift_date(2016, 1, 31, Duration.new!(month: 1))
+      {2016, 2, 29}
+      iex> Calendar.ISO.shift_date(2016, 1, 31, Duration.new!(year: 4, day: 1))
+      {2020, 2, 1}
+  """
+  @spec shift_date(year, month, day, Duration.t()) :: {year, month, day}
+  @impl true
+  def shift_date(year, month, day, duration) do
+    shift_options = get_shift_options(:date, duration)
+
+    {year, month, day, _, _, _, _} =
+      Enum.reduce(shift_options, {year, month, day, 0, 0, 0, {0, 0}}, fn
+        {_, 0}, naive_datetime ->
+          naive_datetime
+
+        {:month, value}, naive_datetime ->
+          shift_months(naive_datetime, value)
+
+        {:day, value}, naive_datetime ->
+          shift_days(naive_datetime, value)
+
+        {time_unit, value}, naive_datetime ->
+          shift_time_unit(naive_datetime, value, time_unit)
```

Comment:
We should raise if any time unit less than day is given to dates. So this clause should not be needed.

Reply:
>  if any time unit less than day is given to dates

Do you mean for the duration? Maybe i am misunderstanding, but imho there would be no harm in supporting:

`Date.shift(Date.utc_today(), second: 86400)`

i wonder if this might even be required depending on the representation of custom interval types people may use 🤔  on the other hand, i see how allowing this invites to be less explicit about what you're doing, so i am happy to be convinced this should not be allowed

Reply:
The issue is what happens when you increment by 1000 seconds? Would someone expect that adding `1000` and then `85400` would change the day? It opens up all kinds of questions and quirks. It should be the opposite: if we want to add this, the use case needs to be justified with actual examples. Until then, it is better to do less.

Reply:
👍 that does convince me, thanks José!